### DISCUSSION TITLE
How to Read Pump Discharge?

### DISCUSSION THREAD

**Discussion by hawklorry** [DHI/mikeio1d#16] [Q&A] [answered]
I'm trying to read pump discharge but the quantity is not given in the res1d object. The quantities list in res1d object are given below. The pump discharge is not listed. Same res1d file given more quantities in MIKE+ (see attached screenshot). I have uploaded the res1d file to the ftp (ftp://tea42@ftp.dhigroup.com/pub/zhyu/richmond_2041OCP_24hr_2.4m_Upgrade2041OCP-24hr-2.4m-UpgradeRichmondHD.res1d) . Thanks for your help. 

['WaterLevel',
 'WaterVolumeAboveGround',
 'WaterFlowRateAboveGround',
 'WaterVolumeRetainedInMaxInflow',
 'Discharge',
 'FlowVelocity',
 'WaterLevelSlope',
 'EnergyLevelSlope']

![DHI MIKEPlus Shell_QOIEhCveF7](https://user-images.githubusercontent.com/3594975/138306093-1eb88994-85a5-4d70-ae29-5df159e825d2.png)

**Comment by gedaskir** [comment, marked as answer]
In this case the provided res1d file does contain only 'Discharge' quantity for structures. Here Mike1D defines the pump on a Q-point of so called structure reach (the same is for weir and orifice). Mike+ plus is interpreting the 'Discharge' quantity of this Q-point as Pump Discharge. Note that there is also other kind of output called 'DischargeInStructure', which is possible to enable in Mike+ (I won't go into detail what is the difference between 'Discharge' and 'DischargeInStructure'):

![DischargeInStructure](https://user-images.githubusercontent.com/10992669/138336760-618b8095-66da-4c62-a2d2-38028b3a9cc0.jpg)

Here is an example script to retrieve a discharge of pump 'Woodward_P4':

```python
from mikeio1d.res1d import Res1D
from mikeio1d.res1d import QueryDataReach

pump_id = "Pump:Woodward_P4"
res1d = Res1D("richmond_2041OCP_24hr_2.4m_Upgrade2041OCP-24hr-2.4m-UpgradeRichmondHD.res1d", reaches=[pump_id])

query_pump = QueryDataReach("Discharge", pump_id, 5.0)
df = res1d.read(query_pump)

print(df)
```
which yields the following output

![PumpDischarge](https://user-images.githubusercontent.com/10992669/138335434-9d038155-65df-42e4-8d8a-918417fa6311.jpg)

I think the mikeio1d API should be improved in this case to retrieve quantities on structures reaches.
For example, now one needs to know that the pump is being put on chainage 5 [m] of the structure reach.

I hope this helps.

**Comment by hawklorry** [comment]
Thanks @gedaskir. So the trick is to add structure type before the name? I got following error. Am I using a different version? 

![Code_EPb429r6BY](https://user-images.githubusercontent.com/3594975/138459764-71ca0ccb-c764-418e-904b-8c7135ab63b0.png)

**Comment by gedaskir** [comment]
Yes, it looks so. For the above script to work install the latest version from main:
``pip install https://github.com/DHI/mikeio1d/archive/main.zip``

The statement *reaches=[pump_id]* allows to load only that time series related to pump. This is more performant.
With your version of mikeio1d try just this:
```python
res1d = Res1D("richmond_2041OCP_24hr_2.4m_Upgrade2041OCP-24hr-2.4m-UpgradeRichmondHD.res1d")
```

**Comment by hawklorry** [comment]
Great. It works now.
